r3bbase/R3BTprevTnext.cxx

```diff
@@ -0,0 +1,170 @@
+#include "R3BShared.h"
+#include "TClonesArray.h"
+#include "TMath.h"
+#include "TObjArray.h"
+#include "TRandom.h"
+#include <cstdlib>
+#include <iostream>
+
+ClassImp(R3BTprevTnext) // NOLINT
+
+    R3BTprevTnext::R3BTprevTnext()
+    : R3BTprevTnext("R3B Tprev/Tnext", 1)
+{
+}
+
+R3BTprevTnext::R3BTprevTnext(const TString& name, Int_t iVerbose)
+    : FairTask(name, iVerbose)
+    , fMSOffsetPar(nullptr)
+    , fTprev(0.0)
+    , fTnext(0.0)
+    , fSamplerMapped(nullptr)
+    , fSamplerMSMapped(nullptr)
+    , fR3BEventHeader(nullptr)
+    , fTpat(1024.0)
+    , delta_clk(1.0)
+{
+}
+
+R3BTprevTnext::~R3BTprevTnext() { LOG(info) << "R3BTprevTnext: Delete instance"; }
+
+InitStatus R3BTprevTnext::Init()
+{
+
+    FairRootManager* rootManager = FairRootManager::Instance();
+    if (rootManager == nullptr)
+    {
+        return kFATAL;
+    }
+
+    FairRuntimeDb* rtdb = FairRuntimeDb::instance();
+    if (rtdb == nullptr)
+    {
+        LOG(error) << "FairRuntimeDb not opened!";
+        return kFATAL;
+    }
+
+    fR3BEventHeader = dynamic_cast<R3BEventHeader*>(rootManager->GetObject("EventHeader."));
+    if (fR3BEventHeader == nullptr)
+    {
+        LOG(error) << "R3BTprevTnext::Init() EventHeader. not found";
+        return kFATAL;
+    }
+
+    fSamplerMapped = dynamic_cast<TClonesArray*>(rootManager->GetObject("SamplerMapped"));
+    if (fSamplerMapped == nullptr)
+    {
+        return kFATAL;
+    }
+
+    fSamplerMSMapped = dynamic_cast<TClonesArray*>(rootManager->GetObject("SamplerMSMapped"));
+    if (fSamplerMSMapped == nullptr)
+    {
+        return kFATAL;
+    }
+
+    fMSOffsetPar = dynamic_cast<R3BMSOffsetPar*>(rtdb->getContainer("MSOffsetPar"));
+    if (fMSOffsetPar == nullptr)
+    {
+        LOG(error) << "Could not find MSOffsetPar container!";
+        return kFATAL;
+    }
+
+    return kSUCCESS;
+}
+
+void R3BTprevTnext::Exec(Option_t* /*opt*/)
+{
+
+    fTprev = -1.0e3; // Initializaton of Tprev and Tnext
+    fTnext = -1.0e3; // If in the end you have this value is because this is an offspill or SAMP and/or SAMPMS are not
+                     // properly recorded
+    const Int_t sampHits = fSamplerMapped->GetEntriesFast();
+    const Int_t sampmsHits = fSamplerMSMapped->GetEntriesFast();
+    if ((sampHits < 1) || (sampmsHits != 1)) // If statement to check if the event is properly recorded
+    {
+        fR3BEventHeader->SetTprev(fTprev);
+        fR3BEventHeader->SetTnext(fTnext);
+        return;
+    }
+    const Double_t MSOffset = fMSOffsetPar->GetMSOffset();
+    Double_t SAMPTime = 0;
+    Double_t SAMPMSTime = 0;
+    R3BSamplerMappedData* SAMPMapped = nullptr;
+    R3BSamplerMappedData* SAMPMSMapped = nullptr;
+
+    SAMPMSMapped = dynamic_cast<R3BSamplerMappedData*>(fSamplerMSMapped->At(0));
+    SAMPMSTime = SAMPMSMapped->GetTime();
+
+    if (fR3BEventHeader->GetTpat() < fTpat && fR3BEventHeader->GetTrigger() == 1 && fR3BEventHeader->GetTpat() > 0)
+    {
+        if (sampHits == 1)
```

Comment:
No need to handle that case specially. Just let the for loop roll over one entry.

Reply:
This one I need to think more about to fully understand